src/librustc_trans_utils/trans_crate.rs

```diff
@@ -73,6 +74,186 @@ pub trait TransCrate {
     ) -> Result<(), CompileIncomplete>;
 }
 
+#[macro_export]
+macro_rules! hot_pluggable_trans_crate {
+    (|$sess:ident| { $body:expr }) => {
+        use $crate::__rustc::ty::maps::Providers;
+        #[no_mangle]
+        pub extern "C" fn __rustc_backend_new($sess: &Session) -> *mut Box<TransCrate> {
```

Comment:
This doesn't need `extern "C"` or to return a `*mut` around `Box<TransCrate>`, there are no ABI implications here.

Also, do you need more functions than this one? That's why I said to add `&self` arguments to all the methods, so you can just call them on a trait object.

Reply:
> This doesn't need extern "C" or to return a *mut around Box<TransCrate>, there are no ABI implications here.

Didnt know that was allowed

Reply:
Well it's no different than the methods of the `TransCrate` trait object, other than being exported by name. The ABI considerations are identical.